File: packages/mcp/src/auth/store.ts
import { mkdir, readFile, rename, rm, writeFile } from "node:fs/promises";
import { join } from "node:path";

export type TokenSet = {
  access_token: string;
  refresh_token?: string;
  token_type?: string;
  expiry: string;
  id_token?: string;
};

export type Session = {
  tokens: TokenSet;
  tenant_id?: string;
  tenant_name?: string;
  subject?: string;
  email?: string;
  name?: string;
  /** Named platform profile used when tokens were issued. */
  profile_id?: string;
  /** OIDC authority / issuer URL at login time. */
  oidc_authority?: string;
  /** Short-lived authoring session for files_* / state local tools. */
  dev_session_id?: string;
  /** Optional agent id used with the authoring session. */
  dev_agent_id?: string;
};

export class SessionStore {
  readonly path: string;

  constructor(stateHome: string) {
    this.path = join(stateHome, "session.json");
  }

  async ensureHome(stateHome: string): Promise<void> {
    await mkdir(stateHome, { recursive: true, mode: 0o700 });
  }

  async load(): Promise<Session | null> {
    try {
      const raw = await readFile(this.path, "utf8");
      return JSON.parse(raw) as Session;
    } catch (err: unknown) {
      if ((err as NodeJS.ErrnoException).code === "ENOENT") return null;
      throw err;
    }
  }

  async save(session: Session): Promise<void> {
    const tmp = `${this.path}.tmp`;
    await writeFile(tmp, JSON.stringify(session, null, 2), { mode: 0o600 });
    await rename(tmp, this.path);
  }

  async clear(): Promise<void> {
    try {
      await rm(this.path);
    } catch (err: unknown) {
      if ((err as NodeJS.ErrnoException).code !== "ENOENT") throw err;
    }
  }
}

File: mcp-server/internal/auth/store.go
package auth

import (
	"encoding/json"
	"errors"
	"os"
	"path/filepath"
	"sync"
	"time"
)

type TokenSet struct {
	AccessToken  string    `json:"access_token"`
	RefreshToken string    `json:"refresh_token,omitempty"`
	TokenType    string    `json:"token_type,omitempty"`
	Expiry       time.Time `json:"expiry"`
	IDToken      string    `json:"id_token,omitempty"`
}

type Session struct {
	Tokens   TokenSet `json:"tokens"`
	TenantID string   `json:"tenant_id,omitempty"`
	Subject  string   `json:"subject,omitempty"`
	Email    string   `json:"email,omitempty"`
	Name     string   `json:"name,omitempty"`
}

type Store struct {
	mu   sync.Mutex
	path string
}

func NewStore(cursorHome string) (*Store, error) {
	if err := os.MkdirAll(cursorHome, 0o700); err != nil {
		return nil, err
	}
	return &Store{path: filepath.Join(cursorHome, "session.json")}, nil
}

func (s *Store) Path() string { return s.path }

func (s *Store) Load() (*Session, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	data, err := os.ReadFile(s.path)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return nil, nil
		}
		return nil, err
	}
	var sess Session
	if err := json.Unmarshal(data, &sess); err != nil {
		return nil, err
	}
	return &sess, nil
}

func (s *Store) Save(sess *Session) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	data, err := json.MarshalIndent(sess, "", "  ")
	if err != nil {
		return err
	}
	tmp := s.path + ".tmp"
	if err := os.WriteFile(tmp, data, 0o600); err != nil {
		return err
	}
	return os.Rename(tmp, s.path)
}

func (s *Store) Clear() error {
	s.mu.Lock()
	defer s.mu.Unlock()
	err := os.Remove(s.path)
	if errors.Is(err, os.ErrNotExist) {
		return nil
	}
	return err
}
